lib/engine/game/g_18_india/step/certificate_selection.rb

```diff
@@ -55,57 +58,75 @@ def players_visible?
             false
           end
 
+          def show_companies
+            true
+          end
+
+          def show_map
+            true
+          end
+
           def name
-            'Hand Selection'
+            'Initial Hand Selection'
           end
 
           def description
             "Select #{@cards_to_keep} Certificates for your starting hand"
           end
 
-          def finished?
-            @confirmed_selections == @game.players.size
+          def selection_note
+            [
+              "Made #{number_of_selections} of #{@cards_to_keep} selections.",
+              'Click on card to select or unselect it.',
+            ]
           end
 
-          def selections_completed?
-            number_of_selections == @cards_to_keep
+          def select_company(player, company)
+            return if selections_completed? && company.owner.nil? # prevent selecting more than allowed
+
+            # add or remove from choices
+            if company_selected?(company)
+              @choices[player].delete(company)
+              company.owner = nil
+            else
+              @choices[player] << company
+              @choices[player].sort!
+              company.owner = player
+            end
           end
 
-          def number_of_selections
-            current_entity.hand.count { |s| s.owner == current_entity }
+          def company_selected?(company)
+            @choices[current_entity].include?(company)
           end
 
-          def actions(entity)
-            return [] if finished?
-            return [] unless entity == current_entity
+          def selected_companies
+            @choices[current_entity]
+          end
 
-            if selections_completed?
-              ACTIONS_WITH_PASS
-            else
-              ACTIONS
-            end
+          def number_of_selections
+            selected_companies.size
           end
 
-          def process_pass(action)
-            @log << "#{action.entity.name} selected #{@cards_to_keep} certificates for hand"
-            @confirmed_selections += 1
-            @round.next_entity_index!
-            action_finalized
+          def selections_completed?
+            number_of_selections == @cards_to_keep
           end
 
-          def process_bid(action)
-            choose_company(action.entity, action.company)
-            @game.next_turn!
+          def process_select_multiple_companies(action)
+            player = action.entity
+            selected_companies = action.companies
```

Comment:
You could be paranoid here and check that all of `selected_companies` are in `player.hand` and throw an error if not.

Reply:
Added a check as a little paranoia can be healthy...